msticpy/data/drivers/local_velociraptor_driver.py

```diff
@@ -0,0 +1,222 @@
+from ...common.provider_settings import get_provider_settings
+from ...common.utility import export, valid_pyname
+from .driver_base import DriverBase, QuerySource
+
+__version__ = VERSION
+__author__ = "ianhelle"
+
+logger = logging.getLogger(__name__)
+
+
+# pylint: disable=too-many-instance-attributes
+@export
+class VelociraptorLogDriver(DriverBase):
+    """OSQueryLogDriver class to execute kql queries."""
+
+    OS_QUERY_DATEIME_COLS = {
+        "unixTime",
+        "columns_time",
+        "columns_atime",
+        "columns_ctime",
+        "columns_mtime",
+    }
+
+    def __init__(self, connection_str: Optional[str] = None, **kwargs):
+        """
+        Instantiate OSQueryLogDriver and optionally connect.
+
+        Parameters
+        ----------
+        connection_str : str, optional
+            Connection string (not used)
+        data_paths : List[str], optional
+            Paths from which to load data files
+        progress : bool, optional
+            Show progress with tqdm, by default, True
+
+        """
+        del connection_str
+        if kwargs.get("debug", False):
+            logger.setLevel(logging.DEBUG)
+        super().__init__()
+
+        self._paths: List[str] = ["."]
+        # If data paths specified, use these
+        # from kwargs or settings
+        if data_paths := kwargs.get("data_paths"):
+            self._paths = [path.strip() for path in data_paths]
+            logger.info("data paths read from param %s", str(self._paths))
+        else:
+            prov_settings = get_provider_settings(config_section="DataProviders").get(
+                "VelociraptorLogs"
+            )
+            if prov_settings:
+                self._paths = prov_settings.args.get("data_paths", []) or self._paths
+                logger.info("data paths read from config %s", str(self._paths))
+
+        self.data_files: Dict[str, List[Path]] = {}
+        self._schema: Dict[str, Any] = {}
+        self._query_map: Dict[str, List[str]]
+        self._progress = kwargs.pop("progress", True)
+        self._loaded = True
+        self.has_driver_queries = True
+        logger.info("data files to read %s", ",".join(self.data_files))
+
+    def connect(self, connection_str: Optional[str] = None, **kwargs):
+        """
+        Connect to data source.
+
+        Parameters
+        ----------
+        connection_str : str
+            Connect to a data source
+
+        """
+        del connection_str
+        self.data_files = self._get_logfile_paths()
+        self._connected = True
+
+    @property
+    def schema(self) -> Dict[str, Dict]:
+        """
+        Return current data schema of connection.
+
+        Returns
+        -------
+        Dict[str, Dict]
+            Data schema of current connection.
+
+        """
+        if not self._schema:
+            if not self.data_files:
+                self.connect()
+            # read the first row of each file to get the schema
```

Comment:
I am not familiar with velocirator, but I assuming that these JSON blobs are not dynamic. Otherwise the schema might miss columns that are not present in the first row of each file.

Reply:
They are or it depends.
Like osquery, the output depends on configured query which is using VQL, a SQL-like format and is customizable by user.
If staying with default artifacts or inside a given collection and artifact, mostly static but not escaping that it can be changed and some edge case like below example logs, likely because of nested data.

See also https://docs.velociraptor.app/docs/vql/artifacts/

Reply:
I'm not sure how expensive we want to make the schema property. Looking at the first record is cheap and cheerful. Even if we did the first 50 or 1000 records, we'd have no guarantee that an extra field was not in record 1001.